narwhal/primary/src/synchronizer.rs

```diff
@@ -74,6 +344,272 @@ impl Synchronizer {
+        let digest = certificate.digest();
+        if self.inner.certificate_store.contains(&digest)? {
+            trace!("Certificate {digest:?} has already been processed. Skip processing.");
+            self.inner.metrics.duplicate_certificates_processed.inc();
+            return Ok(());
+        }
+
+        self.sanitize_certificate(&certificate).await?;
+
+        debug!(
+            "Processing certificate {:?} round:{:?}",
+            certificate,
+            certificate.round()
+        );
+
+        let certificate_source = if self.inner.name.eq(&certificate.origin()) {
+            "own"
+        } else {
+            "other"
+        };
+        let highest_received_round = self
+            .inner
+            .highest_received_round
+            .fetch_max(certificate.round(), Ordering::AcqRel)
+            .max(certificate.round());
+        self.inner
+            .metrics
+            .highest_received_round
+            .with_label_values(&[certificate_source])
+            .set(highest_received_round as i64);
+
+        // Let the proposer draw early conclusions from a certificate at this round and epoch, without its
+        // parents or payload (which we may not have yet).
+        //
+        // Since our certificate is well-signed, it shows a majority of honest signers stand at round r,
+        // so to make a successful proposal, our proposer must use parents at least at round r-1.
+        //
+        // This allows the proposer not to fire proposals at rounds strictly below the certificate we witnessed.
+        let minimal_round_for_parents = certificate.round().saturating_sub(1);
+        self.inner
+            .tx_parents
+            .send((vec![], minimal_round_for_parents, certificate.epoch()))
+            .await
+            .map_err(|_| DagError::ShuttingDown)?;
+
+        // Instruct workers to download any missing batches referenced in this certificate.
+        // Since this header got certified, we are sure that all the data it refers to (ie. its batches and its parents) are available.
+        // We can thus continue the processing of the certificate without blocking on batch synchronization.
+        let inner = self.inner.clone();
+        let header = certificate.header.clone();
+        let sync_network = network.clone();
+        let max_age = self.inner.gc_depth.saturating_sub(1);
+        self.inner.batch_tasks.lock().spawn(async move {
+            Synchronizer::sync_batches_internal(inner, &header, sync_network, max_age).await
+        });
+
+        // Ensure either we have all the ancestors of this certificate, or the parents have been garbage collected.
+        // If we don't, the synchronizer will start fetching missing certificates.
+        if certificate.round() > self.inner.gc_round.load(Ordering::Acquire) + 1
+            && !self.check_parents(&certificate).await?
+        {
+            let rx = {
+                let mut pending = self.inner.pending.lock().await;
```

Comment:
@mwtian do you believe that we should make a similar check in the beginning of the `process_certificate_internal` method? If the certificate is already suspended then we probably want to avoid executing all the previous logic and subscribe directly on the `pending` map.

Reply:
Again I'm mostly keeping the existing logic and leaving improvements to future PRs .. Indeed this is what I will do. The caveat is that if a call to `process_certificate_internal()` awaits on pending certs, there must be logic to accept pending certs when possible i.e. all parents exist. Otherwise no cert that exists in the `pending` map can ever get accepted.

Reply:
I see what you mean and you are right - I thought that the `accept_certificate_internal` would be directly called in this case (when parents are resolved and we are now ok to store the certificate) - but I was wrong on that. Ok let's see the improvements in follow up PR